### PR TITLE
Override local disk state if we are able to restore from remote

### PR DESCRIPTION
**Pull request by linuxpi**
### Description
- Immediately after remote state restore, we write restored state to disk with UNKNOWN_UUID cluster uuid. 
- This leads to issues if process restarts before committing newly generated state with valid uuid to disk
- Now, If restart happens before performing a new write to disk with valid cluster uuid, the auto restore fails with conflicting states in remote and local disk. 
- Please find more details in the linked issue

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/10776

### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java [hunk 1]
```diff
@@ -175,7 +175,10 @@ public void start(
                             if (ClusterState.UNKNOWN_UUID.equals(lastKnownClusterUUID) == false) {
                                 // Load state from remote
                                 final RemoteRestoreResult remoteRestoreResult = remoteStoreRestoreService.restore(
-                                    clusterState,
+                                    // Override Metadata restored from local disk during remote state restore.
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:178]: remove this line it is redundant

File: server/src/main/java/org/opensearch/gateway/GatewayMetaState.java [hunk 2]
```diff
@@ -550,6 +553,7 @@ static class LucenePersistedState implements PersistedState {
             // out by this version of OpenSearch. TODO TBD should we avoid indexing when possible?
             final PersistedClusterStateService.Writer writer = persistedClusterStateService.createWriter();
             try {
+                // With Remote State we can write cluster state with non empty Metadata but UNKNOWN_UUID cluster uuid
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/gateway/GatewayMetaState.java:556]: During remote state restore, there will be non empty metadata getting persisted with cluster UUID as ClusterState.UNKOWN_UUID . The valid UUID will be generated and persisted along with the first cluster state getting published.

File: server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java [hunk 3]
```diff
@@ -295,52 +293,8 @@ private void validate(
...
-                        logger.info(finalErrorMsg);
-                        throw new IllegalStateException(finalErrorMsg);
-                    }
-
-                    boolean isHidden = IndexMetadata.INDEX_HIDDEN_SETTING.get(indexMetadata.getSettings());
-                    createIndexService.validateIndexName(indexName, currentState);
-                    createIndexService.validateDotIndex(indexName, isHidden);
-                    shardLimitValidator.validateShardLimit(indexName, indexMetadata.getSettings(), currentState);
-                } else if (restoreAllShards && IndexMetadata.State.CLOSE.equals(indexMetadata.getState()) == false) {
+                if (restoreAllShards && IndexMetadata.State.CLOSE.equals(indexMetadata.getState()) == false) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java:297]: lets move the respective validations for index restore and metadata restore above itself. This method is not needed anymore.

**Review by Bukhtawar — COMMENTED** [review]
Thanks @linuxpi for the changes, please add an integ test to confirm if the state before and after recovery matches. ALso please populate the PR template to add more context and issue link for the same

**Comment by linuxpi** [comment]
> Thanks @linuxpi for the changes, please add an integ test to confirm if the state before and after recovery matches. ALso please populate the PR template to add more context and issue link for the same

Updated the description with details. Integ tests we already have in place to ensure correctness and recovery is successful. For the particular case we faced this issue we need to introduce a new test. I'm working on writing a test for it.

**Comment by shwetathareja** [comment]
Lets add an IT which covers following sequence of steps
1. Create single node cluster
2. Create an index with documents
3. Stop this node
4. when the new node comes up, change the logic so that it doesn't seed initial_master_nodes
5. restart the new node and seed the initial_master_nodes
6. cluster should come up with index metadata restore and marked red.

**Comment by linuxpi** [comment]
> Lets add an IT which covers following sequence of steps
> 
>     1. Create single node cluster
> 
>     2. Create an index with documents
> 
>     3. Stop this node
> 
>     4. when the new node comes up, change the logic so that it doesn't seed initial_master_nodes
> 
>     5. restart the new node and seed the initial_master_nodes
> 
>     6. cluster should come up with index metadata restore and marked red.

i am working on adding this test by controlling seeding of cluster manager nodes, but currently after all operations, cluster fails to form, i am working in fixing it but doesnt look trivial. 
Let me know if you think we can create a followup issue for this particular test. We have basic integ already in place.

**Review by shwetathareja — APPROVED** [review]

**Comment by linuxpi** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> 
>     * **RESULT:** FAILURE ❌
> 
>     * **URL:** https://build.ci.opensearch.org/job/gradle-check/28723/
> 
>     * **CommitID:** [00cce63](https://github.com/opensearch-project/OpenSearch/commit/00cce63565f354532742b5b83cd869d4f8744039)
>       Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>       Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky Test 

[org.opensearch.cluster.MinimumClusterManagerNodesIT.testThreeNodesNoClusterManagerBlock](https://build.ci.opensearch.org/job/gradle-check/28723/testReport/junit/org.opensearch.cluster/MinimumClusterManagerNodesIT/testThreeNodesNoClusterManagerBlock/)


https://github.com/opensearch-project/OpenSearch/issues/10006

**Pull request merged**